Add rule for missing process block when command supports the pipeline

mattmcnabb opened the pull request:
## PR Summary

This PR adds a rule to warn when a function has at least one parameter that supports the pipeline but doesn't implement a process block. This feature is detailed in issue #1368.

I still need to add tests and docs, but wanted to get this out there to take any feedback on overall approach or general details such as rule naming, messages, etc.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] Make sure you've added a new test if existing tests do not effectively test the code changed and/or updated documentation
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

bergmeister commented:
Awesome, let me know if you need help with writing the test. Is the code for the rule already working and ready for review? In case you encounter issues with the build we are there for you as well, the first goal would be to have at least 1 of the 4 build environments green, I'm happy to help getting the edge cases for other environments right as I've got a bit of experience in it already.
In the meantime, I've re-targeted the PR to the master branch as the development branch is just there for legacy purposes

bergmeister commented:
Awesome, let me know if you need help with writing the test. Is the code for the rule already working and ready for review? In case you encounter issues with the build we are there for you as well, the first goal would be to have at least 1 of the 4 build environments green, I'm happy to help getting the edge cases for other environments right as I've got a bit of experience in it already.
In the meantime, I've re-targeted the PR to the master branch as the development branch is just there for legacy purposes

mattmcnabb commented:
@bergmeister yes the code appears to be working as-is - I just didn’t get a chance to write tests yet. I’ll take a look later today to see if there are any compatibility quirks that trip me up, but I think this one is pretty straight-forward. Thanks!

bergmeister commented:
Because you added strings to the `.resx` files of the rules project the strongly typed classes file needed to be updated to fix the build, I've just pushed a change for that. If you need to change those resource strings again, you can automatically get them re-created by running `.\New-StronglyTypedCsFileForResx.ps1 Rules`

bergmeister commented on Rules/UseProcessBlockForPipelineCommands.cs:
```diff
@@ -0,0 +1,99 @@
+                    && funcAst.Body.ParamBlock.Attributes != null
+                    && funcAst.Body.ParamBlock.Parameters != null
+                    && funcAst.Body.ProcessBlock == null
+                )
```
You have quite a few indentation layers. You can eliminate some of them in some foreach loops by applying the following refactoring that makes the foreach loop continue if a condition is not met to avoid nesting the critical path in an if condition that leads to more indentation
```csharp
foreach(...)
{
    if (condition)
    {
        ...
    }
}
```
becomes
```csharp
foreach(...)
{
    if (!condition)
    {
        continue;
    }
    ...
}
```

mattmcnabb replied:
This should be resolved in dd428297d55f7256d4350133d64161b2679ced68. I took your advice and execution should drop out of each loop if conditions do not match, which keeps the return statement out of the if blocks.

Also, I realized that two of the conditional checks were not necessary:
### Check for null body

```csharp
// before
if
(
        funcAst.Body == null
	|| funcAst.Body.ParamBlock == null
	|| funcAst.Body.ParamBlock.Attributes == null
	|| funcAst.Body.ParamBlock.Parameters == null
	|| funcAst.Body.ProcessBlock != null
)

// after
if
(
	funcAst.Body.ParamBlock == null
	|| funcAst.Body.ParamBlock.Attributes == null
	|| funcAst.Body.ParamBlock.Parameters == null
	|| funcAst.Body.ProcessBlock != null
)
```

 I think checking for a null function body was wrong - I'd still want to warn on a pipeline parameter even if the body hadn't been implemented yet.

### Type check for AttributeAst
```csharp
// removed
if (paramAstAttribute is AttributeAst) { continue; }
```

This check was extraneous.

bergmeister commented on Rules/UseProcessBlockForPipelineCommands.cs:
```diff
@@ -0,0 +1,99 @@
+                            {
+                                var namedArguments = (paramAstAttribute as AttributeAst).NamedArguments;
+                                if (namedArguments != null)
+                                {
```
C# can simplify this logic:
```csharp
                            if (paramAstAttribute is AttributeAst attributeAst)
                            {
                                var namedArguments = attributeAst.NamedArguments;
                                if (attributeAst.NamedArguments != null)
                                {
```

mattmcnabb replied:
Noted, and simplified in dd42829

mattmcnabb commented:
> Because you added strings to the `.resx` files of the rules project the strongly typed classes file needed to be updated to fix the build, I've just pushed a change for that. If you need to change those resource strings again, you can automatically get them re-created by running `.\New-StronglyTypedCsFileForResx.ps1 Rules`

@bergmeister thanks for all your input on this! I'm refactoring based on your style suggestions, but after your change in 1ad72e7 I'm getting the error:

```
Invoke-ScriptAnalyzer : Could not find any resources appropriate for the specified culture or the neutral culture.
Make sure "Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules.Strings.resources" was correctly embedded or
linked into assembly "Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules" at compile time, or that all the
satellite assemblies required are loadable and fully signed.
```

I'm sure I'm missing somethings, but don't see exactly what it is.

bergmeister commented:
Ok, sorry for that, I pushed another commit where I used Visual Studio for the auto-generation, maybe the script doesn't work any more. In the long term view we plan to move to .Net Core 3, which would handle that for us.

bergmeister commented:
I also noticed you had a big diff on the whole resx file, so I reverted the change and re-added your entries for a clean diff. Out of interest, which editor are you using?

mattmcnabb commented:
> I also noticed you had a big diff on the whole resx file, so I reverted the change and re-added your entries for a clean diff in that file. Out of interest, which editor are you using?

using visual studio 2019 - think that was a line endings or encoding issue?

mattmcnabb commented:
I think this is ready to merge now. Let me know what you think about the refactor, tests and the help doc. Again, thanks for your help with this - I'm new to the project so getting started was a bit of a hurdle.

mattmcnabb commented:
Ok, I think that’s all set now. My test are passing, but I’m still seeing several test failures. One test appears to be counting rules, so naturally that failed. 12 others seem to have returned a null reference exception. Not sure about that yet. I’ll leave this be for today and look at it again tomorrow.

mattmcnabb commented:
@bergmeister Well I think I got most of the tests working, but now there is one failing in the Helper tests:

```
Expected the actual value to be less than or equal to 11, because Number of Runspaces should not exceed size of runspace pool cache (10) plus the the default runspace pool (1) in CommandInfoCache, but got 14.

at <ScriptBlock>, C:\projects\psscriptanalyzer\Tests\Engine\Helper.tests.ps1: line 33
33:             (Get-Runspace).Count | Should -BeLessOrEqual 11 -Because 'Number of Runspaces should not exceed size of runspace pool cache (10) plus the the default runspace pool (1) in CommandInfoCache'
```

I'm not sure I can troubleshoot that one, otherwise I think this is ready. Thanks!

bergmeister commented:
That sounds great. You can ignore that one test about runspaces, it is a known, new flaky test, for which we already have a fix in another PR. Thanks so far, I'll try to review the recent changes in the next days

bergmeister commented on RuleDocumentation/UseProcessBlockForPipelineCommands.md:
````diff
@@ -0,0 +1,62 @@
+#### Result
+
+```
+PS C:\> 1..5 | Broken
````
I think you meant this?
```suggestion
PS C:\> 1..5 | Get-Number
```

mattmcnabb replied:
> The only last thing that came to my mind is that maybe `UseProcessBlockForPipelineCommand` might be a better name than `UseProcessBlockForPipelineCommands`, similar to the PowerShell design rule to use single nouns and user's will get a warning for a single command at a time and not commands. I do realize though that some of the other existing rules use a pluralized noun, therefore PSScriptAnalyzer rules already use a mix of styles where it wouldn't matter too much. I'leave that up to your judgement what you think/prefer, especially, since renaming the rule would also require a fair bit of work like renaming the files, etc.

That's fine - I'm always up for any improvements you can suggest. I'm making one more commit with your suggestions and the rule name change so hopefully this is all good now. Thanks again for your help with this, and sorry it took so long to get this up to snuff!

mattmcnabb replied:
Yep, sure did. Good catch! Resolved in 2df5e700a9ba117eed937ff6a715d5f177bd547a.

bergmeister commented on Rules/UseProcessBlockForPipelineCommands.cs:
```diff
@@ -0,0 +1,96 @@
+                            if
+                            (
+                                !String.Equals(namedArgument.ArgumentName, "valuefrompipeline", StringComparison.OrdinalIgnoreCase)
+                                && !String.Equals(namedArgument.ArgumentName, "valuefrompipelinebypropertyname", StringComparison.OrdinalIgnoreCase)
```
```suggestion
                                && !namedArgument.ArgumentName.Equals("valuefrompipelinebypropertyname", StringComparison.OrdinalIgnoreCase)
```

mattmcnabb replied:
fixed in 2df5e700

bergmeister commented on Rules/UseProcessBlockForPipelineCommands.cs:
```diff
@@ -0,0 +1,96 @@
+                {
+                    foreach (var paramAstAttribute in paramAst.Attributes)
+                    {
+                        if (!(paramAstAttribute is AttributeAst)) { continue;  }
```
extraneous space
```suggestion
                        if (!(paramAstAttribute is AttributeAst)) { continue; }
```

mattmcnabb replied:
fixed in 2df5e700

bergmeister commented on Tests/Rules/UseProcessBlockForPipelineCommands.tests.ps1:
```diff
@@ -0,0 +1,37 @@
+        It "has 1 violation for function <Name>" {
+            param ($ScriptDefinition)
+
+            Invoke-ScriptAnalyzer -ScriptDefinition $ScriptDefinition -IncludeRule $RuleName | Should Not BeNullOrEmpty
```
Pester v4 syntax is some backwards compatibility with the v3 (syntax that does not have the dashes) but to stay on the save side, it is better to use the v4 syntax since we use v4.
```suggestion
            Invoke-ScriptAnalyzer -ScriptDefinition $ScriptDefinition -IncludeRule $RuleName | Should -Not -BeNullOrEmpty
```

mattmcnabb replied:
fixed in 2df5e700

bergmeister commented on Tests/Rules/UseProcessBlockForPipelineCommands.tests.ps1:
```diff
@@ -0,0 +1,37 @@
+        It "has no violations for function <Name>" {
+            param ($ScriptDefinition)
+
+            Invoke-ScriptAnalyzer -ScriptDefinition $ScriptDefinition -IncludeRule $RuleName | Should BeNullOrEmpty
```
```suggestion
            Invoke-ScriptAnalyzer -ScriptDefinition $ScriptDefinition -IncludeRule $RuleName | Should -BeNullOrEmpty
```

mattmcnabb replied:
fixed in 2df5e700

mattmcnabb commented on Rules/UseProcessBlockForPipelineCommands.cs:
```diff
@@ -0,0 +1,96 @@
+                        {
+                            if
+                            (
+                                !String.Equals(namedArgument.ArgumentName, "valuefrompipeline", StringComparison.OrdinalIgnoreCase)
```
Thanks, that's a good point. Resolved in 2df5e700

mattmcnabb commented:
> Awesome work, looks good. I only added a couple of minor code style enhancement suggestions that you can apply but otherwise the code looks really good.

Thanks! Just sorry it took so many changes to get up to snuff!

> The only last thing that came to my mind is that maybe `UseProcessBlockForPipelineCommand` might be a better name than `UseProcessBlockForPipelineCommands`, similar to the PowerShell design rule to use single nouns and user's will get a warning for a single command at a time and not commands. I do realize though that some of the other existing rules use a pluralized noun, therefore PSScriptAnalyzer rules already use a mix of styles where it wouldn't matter too much. I'leave that up to your judgement what you think/prefer, especially, since renaming the rule would also require a fair bit of work like renaming the files, etc.

That's fine - I'm always up for any improvements you can suggest. I'm making one more commit with your suggestions and the rule name change so hopefully this is all good now. Find and replace should do the trick!

> P.S. VS 2019 is fine, I use it as well and never had issues editing the resx (just remember to press the save all button so that it updates the `*.designer.cs` file), not sure if it was an encoding issue but at least it's resolved now :-)

Yeah, not sure what happened there. Anyway, since I renamed the command I also modified the strings.resx. This seemed fine and the designer file was updated as well. After I pushed my changes it looks like the entire resx file was diffed, rather than just the changes. That didn't appear to be the case locally prior to pushing. Thoughts?

mattmcnabb commented:
@bergmeister Should I be worried that the rule doesn't highlight the parameter in VSCode:

![image](https://user-images.githubusercontent.com/6998507/69205169-3df76c00-0b17-11ea-808f-a4b2a20230c5.png)

I removed any installations I had of the module and copied the built module into my module path. It appears to be highlighting the existing variable rule, but not mine. The rule does trip on `Invoke-ScriptAnalyzer` though.

bergmeister commented:
In VS-Code not all rules are turned on by default, you can do that either by pointing it to a PSSA settings file or for your purposes just use the `PowerShell: Select ScriptAnalyzer Rules` command, which can enable other rules (but the choice of settings does not persist, i.e. goes back to the default after you close VS-Code)
![image](https://user-images.githubusercontent.com/9250262/69407201-e0704480-0cfb-11ea-99cb-63ddba1e6b74.png)
I made the diff clean again, but I cannot see a difference in encoding or newlines, maybe you have a non-standard VS or git setting. Getting the diff clean is important here because otherwise other people would get an immediate merge conflict in their existing PRs that is not even easy to resolve.

bergmeister commented on Rules/Strings.resx:
```diff
@@ -1080,4 +1080,16 @@
   </data>
-</root>
+  <data name="UseProcessBlockForPipelineCommandCommonName" xml:space="preserve">
+    <value>Use process block for pipeline commands.</value>
```
```suggestion
    <value>Use process block for command that accepts input from pipeline.</value>
```

bergmeister commented on Rules/Strings.resx:
```diff
@@ -1080,4 +1080,16 @@
+    <value>If a parameter accepts input via the pipeline, the function should use at least a process block to correctly support pipeline features.</value>
+  </data>
+  <data name="UseProcessBlockForPipelineCommandError" xml:space="preserve">
+    <value>Commands that support pipeline input should return output in a process block.</value>
```
```suggestion
    <value>Command accepts pipeline input but has not defined a required process block.</value>
```

bergmeister commented on Rules/Strings.resx:
```diff
@@ -1080,4 +1080,16 @@
+    <value>Use process block for pipeline commands.</value>
+  </data>
+  <data name="UseProcessBlockForPipelineCommandDescription" xml:space="preserve">
+    <value>If a parameter accepts input via the pipeline, the function should use at least a process block to correctly support pipeline features.</value>
```
```suggestion
    <value>If a parameter accepts input via the pipeline, the function has to use at least a process block in order to be able to bind the input objects from the pipeline.</value>
```

mattmcnabb commented:
> I made the diff clean again, but I cannot see a difference in encoding or newlines, maybe you have a non-standard VS or git setting. Getting the diff clean is important here because otherwise other people would get an immediate merge conflict in their existing PRs that is not even easy to resolve.

Yeah, not sure what's going on there, but it must be git because it looked fine until I pushed!

rjmholt commented on Rules/Strings.resx:
```diff
@@ -1080,4 +1080,16 @@
+    <value>Use process block for command that accepts input from pipeline.</value>
+  </data>
+  <data name="UseProcessBlockForPipelineCommandDescription" xml:space="preserve">
+    <value>If a parameter accepts input via the pipeline, the function has to use at least a process block in order to be able to bind the input objects from the pipeline.</value>
```
```suggestion
    <value>If a function parameter takes its value from the pipeline, the function must use a process block to bind the input objects from the pipeline to that parameter.</value>
```

rjmholt commented on Rules/Strings.resx:
```diff
@@ -1080,4 +1080,16 @@
+    <value>If a parameter accepts input via the pipeline, the function has to use at least a process block in order to be able to bind the input objects from the pipeline.</value>
+  </data>
+  <data name="UseProcessBlockForPipelineCommandError" xml:space="preserve">
+    <value>Command accepts pipeline input but has not defined a required process block.</value>
```
```suggestion
    <value>Command accepts pipeline input but has not defined a process block.</value>
```

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+﻿using System;
```
Need copyright header

mattmcnabb replied:
fixed in #707f02d96ac5763afaf0b69649d787f8f25b2492

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+    {
+        public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null) throw new ArgumentNullException(Strings.NullAstErrorMessage);
```
```suggestion
            if (ast == null)
            {
                throw new ArgumentNullException(Strings.NullAstErrorMessage);
            }
```

mattmcnabb replied:
fixed in #707f02d96ac5763afaf0b69649d787f8f25b2492

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+        {
+            if (ast == null) throw new ArgumentNullException(Strings.NullAstErrorMessage);
+
+            IEnumerable<Ast> functionAsts = ast.FindAll(testAst => testAst is FunctionDefinitionAst, true);
```
```suggestion
            IEnumerable<Ast> functionAsts = ast.FindAll(testAst => testAst is FunctionDefinitionAst, searchNestedScriptBlocks: true);
```

rjmholt replied:
Currently this applies only the function definitions, but the same should hold for any scriptblock (defined by a `ScriptBlockAst`, which is the type of `FunctionDefinitionAst.Body`). Maybe this rule could be generalised?

mattmcnabb replied:
@rjmholt I think you're right! I'll take a look at that soon to see how easy it would be to implement.

mattmcnabb replied:
Implemented in #707f02d96ac5763afaf0b69649d787f8f25b2492.

I think the only possible change that might need to be made is to update the strings file to reflect "script" or "scriptblock" instead of "function." Thoughts?

rjmholt replied:
I think a simple change like that makes sense. It would also be possible to parameterise the resource string to use a given syntax name and then decide that based on the parent AST type, but there could be various edge cases there and PowerShell doesn't guarantee that the parent AST's type will remain the same for the same syntax across PowerShell versions.

mattmcnabb replied:
Well, i can't seem to push changes to the strings.resx to save my life 🤣 

Even though on commit the diff is just the lines I changes, it ends up diffing the whole file. I've checked my git config and don't see anything that I believe would cause this. I'll keep digging on that.

rjmholt replied:
Are you using Visual Studio? It usually rewrites the whole file. Best to edit the resx in VSCode or another dumber editor

rjmholt replied:
Otherwise it might be encoding or line endings. Not sure what this repo's standard is for either, but should be simple enough to work out

mattmcnabb replied:
Yeah I’m using vs 2019, but when I commit it just appears to show the changes, then when I push the whole file is diffed.

mattmcnabb replied:
I'll try changing again in VS Code. Really the only thing I think I would change replacing the one instance of the work function with "command."

mattmcnabb replied:
So looking at the `New-StronglyTypedCsFileForResx.ps1` script and the comment at the top says:

```powershell
<#
# This script can be used to update the *.Designer file if a *.resx ressource file has been updated.
# However, it is recommended to use Visual Studio instead for editing ressources instead since it takes of that automatically and prodcues cleaner diffs.
#>
```

Weird.

rjmholt replied:
Yeah, so you want to edit the resx in a simple editor, but let Visual Studio do the C# generation from the resx.

VSCode has better encoding/git support, but VS has the tool for resx -> C# conversion.

rjmholt replied:
From the diff this looks like what you've done already. You can't help the massive diff with the Designer.cs file. We basically treat it as a binary since it's autogenerated. You'll see similar diffs in any PRs that touch the resx file in this repo.

mattmcnabb replied:
Ok, so that worked. a7ed1228 has the updated consistent language and didn't diff the whole file.

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+
+            IEnumerable<Ast> functionAsts = ast.FindAll(testAst => testAst is FunctionDefinitionAst, true);
+
+            foreach (FunctionDefinitionAst funcAst in functionAsts)
```
Surprised this compiles without a typecast or a filter

mattmcnabb replied:
@rjmholt Any reason why it wouldn't?

rjmholt replied:
You're supposed to have an `IEnumerable<Ast>` which could contain any `Ast` type (for example, `ExpressionAst`), rather than just `FunctionDefinitionAst`s. So it seems here there is an implicit downcast, which could conceivably cause a runtime error -- seems a bit odd, and I could have sworn I've had to use an explicit cast in this exact case in the past.

mattmcnabb replied:
TBH I find the same usage all over the project, and I think I just robbed it from another rule. However, I have no issue changing it if necessary.

rjmholt replied:
If it compiles then no worries. Just surprised that it does

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+
+            foreach (FunctionDefinitionAst funcAst in functionAsts)
+            {
+                if
```
```suggestion
                if (funcAst.Body.ProcessBlock != null
                    || funcAst.Body.ParamBlock?.Attributes == null
                    || funcAst.Body.ParamBlock.Parameters == null)
                {
                    continue;
                }
```

mattmcnabb replied:
fixed in c7a2531c

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+                {
+                    foreach (var paramAstAttribute in paramAst.Attributes)
+                    {
+                        if (!(paramAstAttribute is AttributeAst)) { continue; }
```
```suggestion
                        if (!(paramAstAttribute is AttributeAst attributeAst)) { continue; }
```

mattmcnabb replied:
resolved in c7a2531c

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+                {
+                    foreach (var paramAstAttribute in paramAst.Attributes)
+                    {
+                        if (!(paramAstAttribute is AttributeAst)) { continue; }
```
```suggestion
                        if (!(paramAstAttribute is AttributeAst paramAttributeAst)) { continue; }
```

mattmcnabb replied:
I think this conflicts with the previous comment

resolved in c7a2531c

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+                        foreach (NamedAttributeArgumentAst namedArgument in namedArguments)
+                        {
+                            if
+                            (
```
Please change if style here to be similar to above (and in other rules)

mattmcnabb replied:
resolved in c7a2531c

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+                    {
+                        if (!(paramAstAttribute is AttributeAst)) { continue; }
+
+                        var namedArguments = (paramAstAttribute as AttributeAst).NamedArguments;
```
With above change, you can eliminate this line

mattmcnabb replied:
resolved in c7a2531c

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,96 @@
+                        if (!(paramAstAttribute is AttributeAst)) { continue; }
+
+                        var namedArguments = (paramAstAttribute as AttributeAst).NamedArguments;
+                        if (namedArguments == null) { continue; }
```
```suggestion
                        if (paramAttributeAst.NamedArguments == null) { continue; }
```

mattmcnabb replied:
resolved in c7a2531c

rjmholt commented on Tests/Rules/UseProcessBlockForPipelineCommand.tests.ps1:
```diff
@@ -0,0 +1,37 @@
+            @{ScriptDefinition = $HasProcessBlock; Name = "HasProcessBlock" }
+            @{ScriptDefinition = $HasProcessBlockByPropertyName; Name = "HasProcessBlockByPropertyName" }
+            @{ScriptDefinition = $NoAttribute; Name = "NoAttribute" }
+			@{ScriptDefinition = $HasTypeDeclaration; Name = "HasTypeDeclaration"}
```
```suggestion
            @{ScriptDefinition = $HasTypeDeclaration; Name = "HasTypeDeclaration"}
```

rjmholt replied:
It looks like tabs have been used here. Please make sure you've converted indents to 4 spaces.

mattmcnabb replied:
this should now be fixed in #707f02d96ac5763afaf0b69649d787f8f25b2492

rjmholt commented on Tests/Rules/UseProcessBlockForPipelineCommand.tests.ps1:
```diff
@@ -0,0 +1,37 @@
+            param ($ScriptDefinition)
+
+            Invoke-ScriptAnalyzer -ScriptDefinition $ScriptDefinition -IncludeRule $RuleName | Should -BeNullOrEmpty
+        } -TestCases $Cases
```
Please move this to before the `It` scriptblock body

mattmcnabb replied:
fixed in #707f02d96ac5763afaf0b69649d787f8f25b2492

rjmholt commented on Tests/Rules/UseProcessBlockForPipelineCommand.tests.ps1:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    Context "When there are no violations" {
+        $Cases = @(
```
Please move this definition to either a `BeforeAll` or to be a literal value in the `TestCases` parameter

mattmcnabb replied:
fixed in #707f02d96ac5763afaf0b69649d787f8f25b2492

Moved the script definitions to the cases array.

mattmcnabb replied:
Also took the opportunity to give the tests more descriptive names.

rjmholt commented on Tests/Rules/UseProcessBlockForPipelineCommand.tests.ps1:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    Context "When there are violations" {
+        $Cases = @(
```
Please move this to the `BeforeAll` or to be a literal value in the `TestCases` parameter

mattmcnabb replied:
fixed in #707f02d96ac5763afaf0b69649d787f8f25b2492

rjmholt commented on Tests/Rules/UseProcessBlockForPipelineCommand.tests.ps1:
```diff
@@ -0,0 +1,37 @@
+            param ($ScriptDefinition)
+
+            Invoke-ScriptAnalyzer -ScriptDefinition $ScriptDefinition -IncludeRule $RuleName | Should -Not -BeNullOrEmpty
+        } -TestCases $Cases
```
Please move this to before the `It` body

mattmcnabb replied:
fixed in #707f02d96ac5763afaf0b69649d787f8f25b2492

rjmholt commented on Tests/Rules/UseProcessBlockForPipelineCommand.tests.ps1:
```diff
@@ -0,0 +1,37 @@
+        $HasProcessBlock = 'function GoodFunc1 { [CmdletBinding()] param ([Parameter(ValueFromPipeline)]$Param1) process { } }'
+        $HasProcessBlockByPropertyName = 'function GoodFunc2 { [CmdletBinding()] param ([Parameter(ValueFromPipelineByPropertyName)]$Param1) process { } }'
+        $NoAttribute = 'function GoodFunc3 { [CmdletBinding()] param ([Parameter()]$Param1) }'
+		$HasTypeDeclaration = 'function GoodFunc3 { [CmdletBinding()] param ([Parameter()][string]$Param1) }'
```
```suggestion
            $HasTypeDeclaration = 'function GoodFunc3 { [CmdletBinding()] param ([Parameter()][string]$Param1) }'
```

mattmcnabb replied:
looks like these were errant tabs - fixed in #707f02d96ac5763afaf0b69649d787f8f25b2492

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,98 @@
+                    continue;
+                }
+                
+                foreach (var paramAst in scriptblockAst.ParamBlock.Parameters)
```
```suggestion
                foreach (ParameterAst paramAst in scriptblockAst.ParamBlock.Parameters)
```

mattmcnabb replied:
fixed in a28b4d25

rjmholt commented on Rules/UseProcessBlockForPipelineCommand.cs:
```diff
@@ -0,0 +1,98 @@
+                
+                foreach (var paramAst in scriptblockAst.ParamBlock.Parameters)
+                {
+                    foreach (var paramAstAttribute in paramAst.Attributes)
```
```suggestion
                    foreach (AttributeBaseAst paramAstAttribute in paramAst.Attributes)
```

mattmcnabb replied:
fixed in a28b4d25

mattmcnabb commented:
@rjmholt @bergmeister just checking in to see if there is anything else you need from me on this PR? Thanks!

bergmeister commented:
From my side it is a yes. If Rob is happy as well with the recent changes that he requested, then I'd merge it.

rjmholt commented:
LGTM!
